### DISCUSSION TITLE
Custom aggregate function in Python

### DISCUSSION THREAD

**Discussion by wnchong** [duckdb/duckdb#3658] [General]
Hi, I have a custom aggregate function (UDF) in Python that I'd like to apply to a GROUP BY query, for example:

```sql
SELECT
    grpid,
    UDF(col1, col2)
FROM tbl
GROUP BY grpid
```
where `UDF` is a custom Python function that accepts arguments of type `pandas.Series` or `numpy.array` and returns a scalar. For example: (this is a trivial example of course)

```python
def UDF(col1, col2):
    return numpy.nanmean(col1 + col2)
```

I understand DuckDB supports map functions in Python, and aggregate functions are supported in C++. Just wondering if it would be easy to enable aggregate function support in Python? 

Thanks! (I love DuckDB!)

EDIT: my use case is traversing a large directory of Parquet files and applying a very heavy UDF on each group. Right now the job takes 12 hours because using Pandas. I'm gradually migrating everything over to DuckDB because it's so much faster, but for group operations I'm still using a for-loop on a pandas.GroupBy iterator.

**Comment by hannes** [comment]
There is no support yet for user defined aggregates from Python.

**Comment by Dutcho** [comment]
I'd like this, too. As workaround, I need to `table.fetchall()` to get a list of all data into Python and then perform the aggregation there.

Hence I'm curious on the status. Any progress in past 2 years? Any concrete outlook?

**Comment by soerenwolfers** [comment]
If duckdb had the option to return result rows in an order where rows belonging to the same group are contiguous (which is cheaper than sorting!), performing aggregation using numba would be easier / faster. 
See https://github.com/duckdb/duckdb/discussions/11092

**Comment by ghost** [comment]
@Dutcho One workaround I use is first applying `array_agg` as in `SELECT py_udf(array_agg(x)) FROM table GROUP BY x", which is easier on memory than loading the full table. I would also welcome real python aggregates though.
